File: src/main/java/org/example/bankramenserver/domain/push/service/GetPushNotificationListService.java
package org.example.bankramenserver.domain.push.service;

import lombok.RequiredArgsConstructor;
import org.example.bankramenserver.domain.push.domain.PushNotification;
import org.example.bankramenserver.domain.push.domain.repository.PushNotificationRepository;
import org.example.bankramenserver.domain.push.presentation.dto.PushNotificationListResponse;
import org.example.bankramenserver.domain.user.facade.UserFacade;
import org.springframework.data.domain.PageRequest;
import org.springframework.stereotype.Service;
import org.springframework.transaction.annotation.Transactional;

import java.time.Clock;
import java.util.List;
import java.util.UUID;

@Service
@RequiredArgsConstructor
public class GetPushNotificationListService {

    private final UserFacade userFacade;
    private final PushNotificationRepository pushNotificationRepository;
    private final Clock clock;

    @Transactional(readOnly = true)
    public PushNotificationListResponse execute(int limit) {
        UUID currentUserId = userFacade.getCurrentUserId();
        List<PushNotification> pushNotifications = pushNotificationRepository.findAllByUser_IdOrderBySentAtDesc(
                currentUserId,
                PageRequest.of(0, limit)
        );
        long unreadCount = pushNotificationRepository.countByUser_IdAndIsReadFalse(currentUserId);

        return PushNotificationListResponse.from(pushNotifications, unreadCount, clock);
    }
}

File: src/main/kotlin/org/example/bankramenserver/domain/push/service/GetPushNotificationListService.kt
package org.example.bankramenserver.domain.push.service

import org.example.bankramenserver.domain.push.domain.repository.PushNotificationRepository
import org.example.bankramenserver.domain.push.presentation.dto.PushNotificationListResponse
import org.example.bankramenserver.domain.user.facade.UserFacade
import org.springframework.data.domain.PageRequest
import org.springframework.stereotype.Service
import org.springframework.transaction.annotation.Transactional
import java.time.Clock

@Service
class GetPushNotificationListService(
    private val userFacade: UserFacade,
    private val pushNotificationRepository: PushNotificationRepository,
    private val clock: Clock,
) {
    @Transactional(readOnly = true)
    fun execute(limit: Int): PushNotificationListResponse {
        val userId = userFacade.currentUserId
        return PushNotificationListResponse.from(
            pushNotificationRepository.findAllByUser_IdOrderBySentAtDesc(userId, PageRequest.of(0, limit)),
            pushNotificationRepository.countByUser_IdAndIsReadFalse(userId),
            clock,
        )
    }
}
